src/dom.js

```diff
@@ -13,35 +13,228 @@
+import select from 'css-select';
+
+// htmlparser2 has a relatively DOM-like tree format, which we'll massage into a DOM elsewhere
+const treeAdapter = require('parse5-htmlparser2-tree-adapter');
+
+const PARSE5_OPTS = {
+  treeAdapter,
+};
 
 /**
- * Parse HTML into a mutable, serializable DOM Document, provided by JSDOM.
- * @private
+ * Parse HTML into a mutable, serializable DOM Document.
+ * The DOM implementation is an htmlparser2 DOM enhanced with basic DOM mutation methods.
  * @param {String} html   HTML to parse into a Document instance
  */
-export function createDocument (html) {
-  const jsdom = new JSDOM(html, {
-    contentType: 'text/html'
-  });
-  const { window } = jsdom;
-  const document = window.document;
-  document.$jsdom = jsdom;
+export function createDocument(html) {
+  const document = parse5.parse(html, PARSE5_OPTS);
+
+  defineProperties(document, DocumentExtensions);
+
+  // Extend Element.prototype with DOM manipulation methods.
+  const scratch = document.createElement('div');
+  // Get a reference to the base Node class - used by createTextNode()
+  document.$$Node = scratch.constructor;
+  const elementProto = Object.getPrototypeOf(scratch);
+  defineProperties(elementProto, ElementExtensions);
+  elementProto.ownerDocument = document;
+
   return document;
 }
+
 /**
  * Serialize a Document to an HTML String
- * @private
  * @param {Document} document   A Document, such as one created via `createDocument()`
  */
-export function serializeDocument (document) {
-  return document.$jsdom.serialize();
+export function serializeDocument(document) {
+  return parse5.serialize(document, PARSE5_OPTS);
 }
 
-/** Like node.textContent, except it works */
-export function setNodeText (node, text) {
-  while (node.lastChild) {
-    node.removeChild(node.lastChild);
+/**
+ * Methods and descriptors to mix into Element.prototype
+ */
+const ElementExtensions = {
+  /** @extends htmlparser2.Element.prototype */
+
+  nodeName: {
+    get() {
+      return this.tagName.toUpperCase();
+    },
+  },
+
+  id: reflectedProperty('id'),
+
+  className: reflectedProperty('class'),
+
+  insertBefore(child, referenceNode) {
+    if (!referenceNode) return this.appendChild(child);
+    treeAdapter.insertBefore(this, child, referenceNode);
+    return child;
+  },
+
+  appendChild(child) {
+    treeAdapter.appendChild(this, child);
+    return child;
+  },
+
+  removeChild(child) {
+    treeAdapter.detachNode(child);
+  },
+
+  remove() {
+    treeAdapter.detachNode(this);
+  },
+
+  textContent: {
+    get() {
+      return this.children.map((node) => node.nodeValue).join('\n');
+    },
+
+    set(text) {
+      this.children = [];
+      treeAdapter.insertText(this, text);
+    },
+  },
+
+  setAttribute(name, value) {
+    if (this.attribs == null) this.attribs = {};
+    if (value == null) value = '';
+    this.attribs[name] = value;
+  },
+
+  removeAttribute(name) {
+    if (this.attribs != null) {
+      delete this.attribs[name];
```

Comment:
Can avoid a shape mutation here:

```suggestion
      this.attribs[name] = null;
```

Reply:
Setting it to null causes errors during serialization. So I'm keeping the original change